open_spiel/python/algorithms/deep_cfr_tf2.py

```diff
@@ -0,0 +1,693 @@
+      # Recompute distribution for numerical errors.
+      probs = strategy
+      probs /= probs.sum()
+      sampled_action = np.random.choice(range(self._num_actions), p=probs)
+      self._add_to_strategy_memory(state.information_state_tensor(other_player),
+          self._iteration, strategy, state.legal_actions_mask(other_player))
+      return self._traverse_game_tree(state.child(sampled_action), player)
+
+  @tf.function
+  def _get_matched_regrets(self, info_state, legal_actions_mask, player):
+    """TF-Graph to calculate regret matching"""
+    advs = self._adv_networks[player]((tf.expand_dims(info_state, axis=0),
+            legal_actions_mask), training=False)[0]
+    advantages = tf.maximum(advs, 0)
+    summed_regret = tf.reduce_sum(advantages)
+    if summed_regret > 0:
+      matched_regrets = advantages / summed_regret
+    else:
+      matched_regrets = tf.one_hot(tf.argmax(
+            tf.where(legal_actions_mask==1, advs, -10e20)), self._num_actions)
+    return advantages, matched_regrets
+
+  def _sample_action_from_advantage(self, state, player):
+    """Returns an info state policy by applying regret-matching.
+
+    Args:
+      state: Current OpenSpiel game state.
+      player: (int) Player index over which to compute regrets.
+    Returns:
+      1. (np-array) Advantage values for info state actions indexed by action.
+      2. (np-array) Matched regrets, prob for actions indexed by action.
+    """
+    info_state = tf.constant(state.information_state_tensor(player), 
+                            dtype=tf.float32)
+    legal_actions_mask = tf.constant(state.legal_actions_mask(player),
+                            dtype=tf.float32)
+    advantages, matched_regrets = self._get_matched_regrets(info_state,
+                            legal_actions_mask, player)
+    return advantages.numpy(), matched_regrets.numpy()
+
+  def action_probabilities(self, state):
+    """Returns action probabilities dict for a single batch."""
+    cur_player = state.current_player()
+    legal_actions = state.legal_actions(cur_player)
+    legal_actions_mask = tf.constant(state.legal_actions_mask(cur_player),
+                            dtype=tf.float32)
+    info_state_vector = tf.constant(state.information_state_tensor(),
+                            dtype=tf.float32)
+    if len(info_state_vector.shape) == 1:
+      info_state_vector = tf.expand_dims(info_state_vector, axis=0)
+    probs = self._policy_network((info_state_vector, legal_actions_mask),
+                            training=False)
+    probs = probs.numpy()
+    return {action: probs[0][action] for action in legal_actions}
+
+  def _get_advantage_dataset(self, player):
+    """returns the collected regrets for the given player as a dataset"""
+    self._advantage_memories[player].shuffle_data()
+    data = tf.data.Dataset.from_tensor_slices(
+            self._advantage_memories[player].data)
+    data = data.shuffle(100000)
```

Comment:
Can you explain this line? What is the hard-coded 100000 doing here? Should this not be a parameter?

Reply:
It is to have different data in one batch if there are multiple epochs in one training iteration. It could also be a parameter but I thought that might make it just more confusing if there is another parameter that can be set, and it shouldn't have a big influence.